File: argus/parse/ctags.py
from __future__ import annotations

import json
import logging
import re
import shutil
import subprocess
from dataclasses import dataclass, field
from pathlib import Path, PurePosixPath

from .filters import HEADER_EXTENSIONS

logger = logging.getLogger(__name__)

PRIVATE_SCOPES = frozenset({"detail", "internal", "impl", "anonymous"})

# Universal Ctags has no internal timeout of its own: one pathological
# input that makes it spin would otherwise block the indexer forever, with
# no error and no way to tell which repo it died on (repo_time_budget_seconds
# is only checked in the per-file loop, never once ctags has been invoked).
# 600s is generous even for a large real batch of files in one repo.
CTAGS_TIMEOUT_SECONDS = 600

CTAGS_ARGS = [
    "--output-format=json",
    # n=line, K=long kind, S=signature, s=scope, e=end line,
    # f=file-limited visibility (i.e. `static`), surfaced as JSON key "file".
    "--fields=+nKSsef",
    # Universal Ctags disables the `prototype` kind by default for C/C++;
    # without it, header-only declarations (e.g. `int Foo(int);`) are
    # dropped entirely rather than reported with kind "prototype".
    "--kinds-c=+p",
    "--kinds-c++=+p",
    "-L", "-",   # read the file list from stdin
    "-f", "-",   # write tags to stdout
]


class CtagsUnavailable(RuntimeError):
    """universal-ctags is not installed or is the wrong implementation."""


@dataclass(frozen=True)
class SymbolBatch:
    """The result of one ctags invocation, including what it did NOT process.

    The symbol map alone cannot answer "was this path processed?". ctags
    emits no output at all for a file it parsed cleanly but that contains no
    taggable declarations (an include-only .c, a macro-only header), which is
    byte-for-byte indistinguishable from a file it never opened. Callers that
    stamp a per-path completion marker need that distinction or they will
    mark uncovered files complete forever, so it is reported explicitly:
    every path handed to extract_symbols lands in exactly one of `covered`
    and `uncovered`.
    """

    symbols: dict[str, list[dict]] = field(default_factory=dict)
    covered: frozenset[str] = frozenset()
    uncovered: dict[str, str] = field(default_factory=dict)   # path -> why


def _paths_named_in(paths: list[str], stderr: str) -> set[str]:
    """The listed paths ctags explicitly complained about.

    ctags names the offending file in its diagnostics ("cannot open input
    file \"x.c\""), which is the only attribution available: the JSON output
    format has no per-file markers.

    Matching is NOT a plain substring test: `main.c` and `sub/main.c` are
    ordinary namesakes in real C repos, and a naive `p in haystack` blames
    the root `main.c` for a diagnostic that only ever named `sub/main.c`
    (any path that is a substring of another listed path is at risk). That
    is not a safe over-match -- being blamed is destructive AND budgeted:
    the caller deletes the symbol rows ctags just extracted for the
    wrongly-blamed path, NULLs its symbols_sha, and charges it a retry
    attempt via `_cap_retries`. Three such passes permanently strand a
    perfectly healthy file as symbol-less and retry-exhausted.

    Each candidate path is therefore matched only at a path boundary: the
    character immediately before and after the match (if any) must not
    itself be a path-continuation character (word char, `.`, `/`, `-`).
    This handles both quoted diagnostics ("cannot open input file
    \"sub/main.c\"") and unquoted ones (ctags does not always quote the
    path) without needing separate quote-parsing logic, since `/` counts
    as a path-continuation character on the left: `main.c` inside
    `sub/main.c` is always preceded by `/`, which fails the boundary check.
    """
    if not stderr:
        return set()
    haystack = stderr.replace("\\", "/")
    blamed = set()
    for p in paths:
        pattern = r"(?<![\w./-])" + re.escape(p) + r"(?![\w./-])"
        if re.search(pattern, haystack):
            blamed.add(p)
    return blamed


def is_public_symbol(path: str, scope: str | None, file_restricted: bool) -> bool:
    if scope:
        parts = {p.strip() for p in scope.replace("::", ".").split(".")}
        # "__anon..." is an undocumented ctags-internal naming convention for
        # anonymous-namespace scopes, not a stable public API — a future
        # ctags release changing it would silently reclassify anonymous-
        # namespace symbols as public.
        if any(p in PRIVATE_SCOPES or p.startswith("__anon") for p in parts):
            return False
    if PurePosixPath(path).suffix.lower() in HEADER_EXTENSIONS:
        return True
    return not file_restricted


def extract_symbols(root: Path, rel_paths: list[str]) -> SymbolBatch:
    """Run ctags over rel_paths (relative to root).

    Returns a SymbolBatch: the extracted symbols plus an explicit account of
    which of rel_paths were genuinely processed. Callers must not infer
    coverage from the symbol map (see SymbolBatch).

    A repo-global failure (ctags missing, wrong implementation, timed out)
    still raises CtagsUnavailable -- that is not a per-path outcome.
    """
    if not rel_paths:
        return SymbolBatch()
    exe = shutil.which("ctags")
    if exe is None:
        raise CtagsUnavailable(
            "ctags not found on PATH — install universal-ctags on the index host"
        )

    # A path git checked out that the filesystem will not hand back (a
    # Windows reserved name like aux.c, a >260-char path, a trailing dot) is
    # dropped from the argument list here. It was never processed, so it is
    # uncovered -- not "processed, found nothing".
    existing = [p for p in rel_paths if (root / p).is_file()]
    listed = set(existing)
    uncovered = {
        p: "not a readable file on disk when ctags ran"
        for p in rel_paths if p not in listed
    }
    if not existing:
        return SymbolBatch(uncovered=uncovered)

    try:
        proc = subprocess.run(
            [exe, *CTAGS_ARGS],
            input="\n".join(existing),
            cwd=root,
            capture_output=True,
            text=True,
            encoding="utf-8",
            errors="replace",
            timeout=CTAGS_TIMEOUT_SECONDS,
        )
    except subprocess.TimeoutExpired as exc:
        raise CtagsUnavailable(
            f"ctags timed out after {CTAGS_TIMEOUT_SECONDS}s"
        ) from exc
    if proc.returncode != 0 and not proc.stdout:
        raise CtagsUnavailable(f"ctags failed: {proc.stderr.strip()[:500]}")

    results: dict[str, list[dict]] = {}
    for line in proc.stdout.splitlines():
        if not line.strip():
            continue
        try:
            entry = json.loads(line)
        except json.JSONDecodeError:
            continue
        if entry.get("_type") != "tag":
            continue

        path = PurePosixPath(entry["path"].replace("\\", "/")).as_posix()
        scope = entry.get("scope")
        results.setdefault(path, []).append({
            "name": entry["name"],
            "kind": entry.get("kind", "unknown"),
            "line": int(entry.get("line", 0)),
            "end_line": int(entry["end"]) if entry.get("end") else None,
            "signature": entry.get("signature"),
            "scope": scope,
            "is_public": int(is_public_symbol(path, scope, bool(entry.get("file", False)))),
        })

    if proc.returncode == 0:
        return SymbolBatch(results, frozenset(existing), uncovered)

    # Partial batch: some tags came back, but ctags also reported a non-zero
    # exit. There is no conn here to record it against a repo, so at least
    # surface it in the logs instead of silently discarding proc.stderr.
    detail = proc.stderr.strip()[:500]
    logger.warning(
        "ctags exited %s with partial output for %d file(s): %s",
        proc.returncode, len(existing), detail,
    )
    blamed = _paths_named_in(existing, proc.stderr)
    # When ctags named the files it choked on, everything else in the batch
    # really was processed -- including the symbol-free ones, which is the
    # whole reason coverage is tracked separately from the symbol map.
    # When it named nothing, no such claim can be made, so only paths that
    # actually produced tags count as covered.
    covered = listed - blamed if blamed else set(results) & listed
    for path in sorted(listed - covered):
        uncovered[path] = (
            f"ctags exited {proc.returncode} without covering this file: "
            f"{detail or 'no diagnostic output'}"
        )
    return SymbolBatch(results, frozenset(covered), uncovered)

File: src/Argus/Indexing/Ctags.cs
using System.Diagnostics;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;
using Argus.Store;
using Argus.Util;

namespace Argus.Indexing;

/// <summary>universal-ctags is not installed, is the wrong implementation, or timed out.</summary>
public sealed class CtagsUnavailable(string message) : Exception(message);

/// <summary>
/// The result of one ctags invocation, including what it did NOT process.
/// Every path handed in lands in exactly one of <see cref="Covered"/> and
/// <see cref="Uncovered"/>; the symbol map
/// alone cannot answer "was this path processed".
/// </summary>
public sealed record SymbolBatch(
    Dictionary<string, List<Writes.SymbolRow>> Symbols,
    HashSet<string> Covered,
    Dictionary<string, string> Uncovered,
    HashSet<string> Unattributable)
{
    public static SymbolBatch Empty() => new(new(StringComparer.Ordinal), new(StringComparer.Ordinal), new(StringComparer.Ordinal), new(StringComparer.Ordinal));
}

/// <summary>Symbol extraction with universal-ctags.</summary>
public static class Ctags
{
    public static readonly HashSet<string> PrivateScopes = new(StringComparer.Ordinal) { "detail", "internal", "impl", "anonymous" };
    public const int TimeoutSeconds = 600;

    /// <summary>Bump when a symbol ROW means something different; composed into the per-file stamp.</summary>
    public const string SymbolContractVersion = "3";

    public static readonly string[] Args =
    [
        "--output-format=json",
        "--fields=+nKSsefl", // l: the language, which decides whether a docstring is read
        "--kinds-c=+p",
        "--kinds-c++=+p",
        "-L", "-",
        "-f", "-",
    ];

    /// <summary>Locate an executable on PATH, as shutil.which does.</summary>
    public static string? Which(string name)
    {
        var path = Environment.GetEnvironmentVariable("PATH") ?? "";
        var exts = OperatingSystem.IsWindows()
            ? (Environment.GetEnvironmentVariable("PATHEXT") ?? ".EXE;.CMD;.BAT").Split(';')
            : [""];
        foreach (var dir in path.Split(Path.PathSeparator, StringSplitOptions.RemoveEmptyEntries))
            foreach (var ext in exts)
            {
                var candidate = Path.Combine(dir, name + ext);
                if (File.Exists(candidate)) return candidate;
            }
        return null;
    }

    static HashSet<string> PathsNamedIn(IEnumerable<string> paths, string stderr)
    {
        var blamed = new HashSet<string>(StringComparer.Ordinal);
        if (string.IsNullOrEmpty(stderr)) return blamed;
        var haystack = stderr.Replace('\\', '/');
        foreach (var p in paths)
        {
            var pattern = @"(?<![\w./-])" + Regex.Escape(p) + @"(?![\w./-])";
            if (Regex.IsMatch(haystack, pattern, RegexOptions.CultureInvariant)) blamed.Add(p);
        }
        return blamed;
    }

    public static bool IsPublicSymbol(string path, string? scope, bool fileRestricted)
    {
        if (!string.IsNullOrEmpty(scope))
        {
            var parts = scope.Replace("::", ".").Split('.').Select(p => PyStr.Strip(p));
            if (parts.Any(p => PrivateScopes.Contains(p) || p.StartsWith("__anon", StringComparison.Ordinal))) return false;
        }
        if (Filters.HeaderExtensions.Contains(Filters.Suffix(path).ToLowerInvariant())) return true;
        return !fileRestricted;
    }

    /// <summary>Run ctags over <paramref name="relPaths"/> (relative to <paramref name="root"/>).</summary>
    public static SymbolBatch ExtractSymbols(string root, IReadOnlyList<string> relPaths)
    {
        if (relPaths.Count == 0) return SymbolBatch.Empty();
        var exe = Which("ctags") ?? throw new CtagsUnavailable(
            "ctags not found on PATH — install universal-ctags on the index host");

        var existing = relPaths.Where(p => File.Exists(Path.Combine(root, p))).ToList();
        var listed = new HashSet<string>(existing, StringComparer.Ordinal);
        var uncovered = new Dictionary<string, string>(StringComparer.Ordinal);
        foreach (var p in relPaths)
            if (!listed.Contains(p)) uncovered[p] = "not a readable file on disk when ctags ran";
        if (existing.Count == 0)
            return new SymbolBatch(new(StringComparer.Ordinal), new(StringComparer.Ordinal), uncovered, new(StringComparer.Ordinal));

        var psi = new ProcessStartInfo(exe)
        {
            WorkingDirectory = root,
            RedirectStandardInput = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            StandardOutputEncoding = new UTF8Encoding(false),
            StandardErrorEncoding = new UTF8Encoding(false),
            StandardInputEncoding = new UTF8Encoding(false),
        };
        foreach (var a in Args) psi.ArgumentList.Add(a);

        string stdout, stderr;
        int exitCode;
        using (var proc = Process.Start(psi) ?? throw new CtagsUnavailable("ctags failed: could not start process"))
        {
            var outTask = proc.StandardOutput.ReadToEndAsync();
            var errTask = proc.StandardError.ReadToEndAsync();
            proc.StandardInput.Write(string.Join("\n", existing));
            proc.StandardInput.Close();
            if (!proc.WaitForExit(TimeSpan.FromSeconds(TimeoutSeconds)))
            {
                try { proc.Kill(entireProcessTree: true); } catch (InvalidOperationException) { }
                throw new CtagsUnavailable($"ctags timed out after {TimeoutSeconds}s");
            }
            stdout = outTask.GetAwaiter().GetResult();
            stderr = errTask.GetAwaiter().GetResult();
            exitCode = proc.ExitCode;
        }
        if (exitCode != 0 && stdout.Length == 0)
            throw new CtagsUnavailable($"ctags failed: {PyStr.Prefix(PyStr.Strip(stderr), 500)}");

        var results = new Dictionary<string, List<Writes.SymbolRow>>(StringComparer.Ordinal);
        foreach (var line in PyStr.SplitLines(stdout))
        {
            if (PyStr.Strip(line).Length == 0) continue;
            JsonElement entry;
            try { entry = JsonDocument.Parse(line).RootElement; }
            catch (JsonException) { continue; }
            if (entry.ValueKind != JsonValueKind.Object) continue;
            if (!entry.TryGetProperty("_type", out var type) || type.GetString() != "tag") continue;

            var path = NormalisePath(Str(entry, "path") ?? "");
            var scope = Str(entry, "scope");
            long? end = entry.TryGetProperty("end", out var e) && Truthy(e) ? ToLong(e) : null;
            bool fileRestricted = entry.TryGetProperty("file", out var f) && Truthy(f);
            if (!results.TryGetValue(path, out var list)) results[path] = list = [];
            list.Add(new Writes.SymbolRow(
                Name: Str(entry, "name") ?? "",
                Kind: Str(entry, "kind") ?? "unknown",
                Line: entry.TryGetProperty("line", out var l) ? ToLong(l) : 0,
                EndLine: end,
                Signature: Str(entry, "signature"),
                Scope: scope,
                IsPublic: IsPublicSymbol(path, scope, fileRestricted) ? 1 : 0,
                Doc: null,
                Language: Str(entry, "language") is { Length: > 0 } lang ? lang : null));
        }

        if (exitCode == 0)
            return new SymbolBatch(results, listed, uncovered, new(StringComparer.Ordinal));

        var detail = PyStr.Prefix(PyStr.Strip(stderr), 500);
        Console.Error.WriteLine($"ctags exited {exitCode} with partial output for {existing.Count} file(s): {detail}");
        var blamed = PathsNamedIn(existing, stderr);
        var covered = blamed.Count > 0
            ? new HashSet<string>(listed.Where(p => !blamed.Contains(p)), StringComparer.Ordinal)
            : new HashSet<string>(results.Keys.Where(listed.Contains), StringComparer.Ordinal);
        var misses = listed.Where(p => !covered.Contains(p)).OrderBy(p => p, StringComparer.Ordinal).ToList();
        foreach (var p in misses)
            uncovered[p] = $"ctags exited {exitCode} without covering this file: {(detail.Length > 0 ? detail : "no diagnostic output")}";
        var unattributable = blamed.Count == 0 ? new HashSet<string>(misses, StringComparer.Ordinal) : new HashSet<string>(StringComparer.Ordinal);
        return new SymbolBatch(results, covered, uncovered, unattributable);
    }

    /// <summary>PurePosixPath(p.replace('\\','/')).as_posix(): collapse duplicate and trailing slashes and "." parts.</summary>
    static string NormalisePath(string raw)
    {
        var p = raw.Replace('\\', '/');
        bool abs = p.StartsWith('/');
        var parts = p.Split('/').Where(x => x.Length > 0 && x != ".");
        var joined = string.Join("/", parts);
        if (abs) joined = (p.StartsWith("//") && !p.StartsWith("///") ? "//" : "/") + joined;
        return joined.Length == 0 ? "." : joined;
    }

    static string? Str(JsonElement e, string name) =>
        e.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;

    static bool Truthy(JsonElement v) => v.ValueKind switch
    {
        JsonValueKind.True => true,
        JsonValueKind.Number => v.GetDouble() != 0,
        JsonValueKind.String => v.GetString()!.Length > 0,
        JsonValueKind.Array => v.GetArrayLength() > 0,
        JsonValueKind.Object => v.EnumerateObject().Any(),
        _ => false,
    };

    static long ToLong(JsonElement v) => v.ValueKind switch
    {
        JsonValueKind.Number => v.TryGetInt64(out var n) ? n : (long)v.GetDouble(),
        JsonValueKind.String => long.Parse(v.GetString()!),
        _ => 0,
    };
}
